core/services/form/meta/Attributes.php

```diff
@@ -0,0 +1,198 @@
+    ];
+
+
+    /**
+     * Attributes constructor.
+     *
+     * @param JsonDataHandler $json_data_handler
+     * @param InputTypes      $input_types
+     * @param array           $attributes
+     */
+    public function __construct(JsonDataHandler $json_data_handler, array $attributes, InputTypes $input_types)
+    {
+        $this->json_data_handler = $json_data_handler;
+        $this->input_types = $input_types;
+        $this->setAttributes($attributes);
+    }
+
+
+    /**
+     * @param string $json
+     * @return Attributes
+     */
+    public static function fromJson(string $json): Attributes
+    {
+        $json_data_handler = new JsonDataHandler();
+        $json_data_handler->configure(JsonDataHandler::DATA_TYPE_OBJECT);
+        $attributes = $json_data_handler->decodeJson($json);
+        return LoaderFactory::getNew(Attributes::class, [ $json_data_handler, $attributes ]);
+    }
+
+
+    /**
+     * @return array
+     */
+    public function toArray(): array
+    {
+        return $this->attributes();
+    }
+
+
+    /**
+     * @return string
+     */
+    public function toJson(): string
+    {
+        return $this->json_data_handler->encodeData($this->attributes());
+    }
+
+
+    /**
+     * @param string                $attribute
+     * @param bool|float|int|string $value
+     * @return bool|float|int|string
+     */
+    private function sanitize(string $attribute, $value)
+    {
+        if ($attribute === 'type') {
+            $valid_types = $this->input_types->validTypeOptions();
+            return in_array($value, $valid_types, true) ?: Text::TYPE_TEXT;
+        }
+        $type = $this->attribute_types[ $attribute ] ?? 'string';
+        switch ($type) {
+            case 'bool':
+                return filter_var($value, FILTER_VALIDATE_BOOLEAN);
+            case 'int':
+                return filter_var($value, FILTER_SANITIZE_NUMBER_INT);
+            case 'float':
+                return filter_var($value, FILTER_SANITIZE_NUMBER_FLOAT, FILTER_FLAG_ALLOW_FRACTION);
+            case 'string':
+            default:
+                return filter_var(
+                    $value,
+                    FILTER_SANITIZE_STRING,
+                    FILTER_FLAG_ENCODE_LOW | FILTER_FLAG_ENCODE_HIGH | FILTER_FLAG_ENCODE_AMP
+                );
+        }
+    }
+
+
+    /**
+     * Custom HTML classes to be applied to this form input's help text.
+     * returns a concatenated string unless $as_array is set to true
+     *
+     * @return array
+     */
+    public function attributes(): array
+    {
+        return $this->attributes;
+    }
+
+
+    /**
+     * @param string                $attribute
+     * @param bool|float|int|string $value
+     */
+    public function addAttribute(string $attribute, $value): void
+    {
+        if (array_key_exists($attribute, $this->attributes)) {
+            $this->attributes[ $attribute ] = $this->sanitize($attribute, $value);
+        }
```

Comment:
Is this `array_key_exists` check correct?
Wouldn't it mean that an attribute will never be added if it's not already there?

Reply:
oops ya good catch... thnx for fixing 👍🏻

Reply:
oops... ya you are correct...
what i **_MEANT_** to do was check if the key exists in the `attribute_types` array:

```
if (array_key_exists($attribute, $this->attribute_types)) {
```

which then essentially operates as an `allow/deny` list, meaning an attribute can ONLY be added if it is hardcoded in the `attribute_types` array.